build.py

```diff
@@ -13,48 +13,93 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-# Rebuild samples whose configurations are specified on cli
+# Rebuild samples whose configurations are specified on cli. Runs all
+# requested builds in parallel.
+#
 # Example:
 #     ./build.py config/twemoji-*
 
+import os
 from pathlib import Path
 import shutil
 import subprocess
 import sys
 import time
 
+
 def main():
-    build_dir = Path("build")
-    font_dir = Path("fonts")
+    self_dir = Path(__file__).parent
+    build_dir = Path("build").resolve()
+    if build_dir.exists():
+        shutil.rmtree(build_dir)
+
+    font_dir = Path("fonts").resolve()
+    config_dir = self_dir / "config"
+    font_srcs_dir = self_dir / "font-srcs"
+
+    assert config_dir.is_dir()
+    assert font_srcs_dir.is_dir()
+
+    run_me = []
+
+    setup_start = time.monotonic()
     for config in sys.argv[1:]:
-        config = Path(config)
-        if config.suffix == '.toml':
-            cmd = (
-                "nanoemoji",
-                "--config",
-                str(config)
-            )
-        elif config.suffix == '.py':
-            cmd = (
-                sys.executable,
-                str(config.resolve()),
-                str(build_dir.resolve())
-            )
+        config = Path(config).resolve()
+        working_dir = self_dir / "build" / config.stem
+        working_dir.mkdir(parents=True)
+
+        os.symlink(config_dir, working_dir / config_dir.name)
+        os.symlink(font_srcs_dir, working_dir / font_srcs_dir.name)
+
+        if config.suffix == ".toml":
+            cmd = ("nanoemoji", "--config", str(config))
+        elif config.suffix == ".py":
+            cmd = (sys.executable, str(config.resolve()), str(build_dir.resolve()))
         else:
-            raise ValueError(f'Not sure how to handle {config}')
+            raise ValueError(f"Not sure how to handle {config}")
+
+        run_me.append((working_dir, cmd))
+
+    spawn_start = time.monotonic()
+    procs = [
+        subprocess.Popen(
+            cmd,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            cwd=working_dir,
+            text=True,
+        )
+        for working_dir, cmd in run_me
+    ]
+    for proc, (working_dir, cmd) in zip(procs, run_me):
+        stdout, stderr = proc.communicate()
```

Comment:
like this https://github.com/googlefonts/color-fonts/pull/76

> I'm really interested in when all are done

hm right.. But say the first process takes ages, and the subsequent ones finish earlier, it'd be still nice to be able to see in the main thread that collects results and prints stuff that they have suceeded or failed.
Also with a pool you can manage multiple threads that in turn control subprocesses more efficiently, in that as soon as one completes it makes room for another one within the limits of the max_workers that you set up.

Reply:
> I think we'd have to propagate limits down to nanoemoji for it to truly work.

hm maybe, is it because each separate ninja would not know that another one is running at the same time and will try to parallelize as much as it can, clogging the cpu?
 
I think we don't want to run as many nanoemojis subprocesses as there are os.cpu_count() because then ninja would not have any more room to parallelize in turn within each nanoemoji parallele run.
But if we do, say, cpu_count() // 3 or so, then each ninja could still be able to do stuff in parallel... or not.

Reply:
I think it'd still be better to generate a meta build.ninja and let a single ninja do its thing

Reply:
> because each separate ninja would not know that another one is running

Exactly. You launch, for napkin math purposes, 50 ninja's, some of them are small jobs but 10-20 of them want to run 3k picosvgs, and they don't really stay out of each others way.

Hm. Actually ... I wonder if we couldn't avoid everyone running picosvg. Ninja tries to avoid rework so if we ran all the picosvgs and implanted them inot the filesystem the ninja jobs see they would *not* run picosvg and just run the slow singlethreaded bit. In that case on a modern computer you wouldn't necessarily need any further limiting.

Reply:
> I think it'd still be better to generate a meta build.ninja and let a single ninja do its thing

It would be, but only if you actually put all the individual tasks in so it runs the whole graph, shares picosvgs, etc.  If you simply write one ninja that says "run 50 other ninjas" the value is low. Probably that means nanoemoji itself has to take multiple configs. Not a big deal but certainly more work than what I did :)

That's more work so I didn't do it :) What I've got does quite effectively allow me to do things like "build all the noto fonts again" and have them execute in parallel.